File: lib/player/cdn_token_headers.dart
import 'dart:convert';

/// Azam / Nagra CDN URLs embed a signed `tok_<jwt>` segment. The JWT payload
/// includes a `url` field that must be sent as [Referer] and [Origin].
class CdnTokenHeaders {
  CdnTokenHeaders._();

  static final _tokPattern = RegExp(r'/tok_([^.]+)\.([^.]+)\.([^/]+)/');

  /// Returns `{referer, origin}` when [rawUrl] contains a parsable CDN token.
  static Map<String, String>? refererOriginForUrl(String rawUrl) {
    final match = _tokPattern.firstMatch(rawUrl.trim());
    if (match == null) return null;
    final payloadSegment = match.group(2);
    if (payloadSegment == null || payloadSegment.isEmpty) return null;
    try {
      final bytes = base64Url.decode(base64Url.normalize(payloadSegment));
      final decoded = jsonDecode(utf8.decode(bytes));
      if (decoded is! Map) return null;
      final allowed = (decoded['url'] ?? decoded['referer'] ?? decoded['origin'])
          ?.toString()
          .trim();
      if (allowed == null || allowed.isEmpty) return null;
      final uri = Uri.tryParse(allowed);
      if (uri == null || !uri.hasScheme || uri.host.isEmpty) return null;
      final port = (uri.hasPort && uri.port != 80 && uri.port != 443)
          ? ':${uri.port}'
          : '';
      final origin = '${uri.scheme}://${uri.host}$port';
      final referer = origin.endsWith('/') ? origin : '$origin/';
      return {'referer': referer, 'origin': origin};
    } catch (_) {
      return null;
    }
  }

  static bool isTokenizedCdnUrl(String? rawUrl) {
    if (rawUrl == null || rawUrl.trim().isEmpty) return false;
    return _tokPattern.hasMatch(rawUrl.trim());
  }
}

File: android/app/src/main/kotlin/com/ayubu/supasoka/player/CdnTokenHeaders.kt
package com.ayubu.supasoka.player

import android.util.Base64
import org.json.JSONObject

/** Azam/Nagra `tok_<jwt>` URLs embed allowed Referer/Origin in the JWT payload `url` field. */
object CdnTokenHeaders {
    private val tokPattern = Regex("""/tok_([^.]+)\.([^.]+)\.([^/]+)/""")

    fun refererOriginForUrl(rawUrl: String): Pair<String, String>? {
        val match = tokPattern.find(rawUrl.trim()) ?: return null
        val payloadSegment = match.groupValues.getOrNull(2)?.trim().orEmpty()
        if (payloadSegment.isEmpty()) return null
        return try {
            val jsonText = String(
                Base64.decode(
                    payloadSegment.replace('-', '+').replace('_', '/'),
                    Base64.DEFAULT,
                ),
                Charsets.UTF_8,
            )
            val payload = JSONObject(jsonText)
            val allowed = payload.optString("url")
                .ifBlank { payload.optString("referer") }
                .ifBlank { payload.optString("origin") }
                .trim()
            if (allowed.isEmpty()) return null
            val uri = android.net.Uri.parse(allowed)
            if (uri.scheme.isNullOrBlank() || uri.host.isNullOrBlank()) return null
            val port = when (uri.port) {
                -1, 80, 443 -> ""
                else -> ":${uri.port}"
            }
            val origin = "${uri.scheme}://${uri.host}$port"
            val referer = if (origin.endsWith("/")) origin else "$origin/"
            referer to origin
        } catch (_: Exception) {
            null
        }
    }

    fun isTokenizedCdnUrl(rawUrl: String?): Boolean {
        if (rawUrl.isNullOrBlank()) return false
        return tokPattern.containsMatchIn(rawUrl.trim())
    }
}
